crates/task-impls/src/quorum_proposal.rs

```diff
@@ -461,83 +448,111 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>> QuorumProposalTaskState<TYPE
             event_receiver,
         );
 
+        info!(
+            "Node {} Dependency {:?} is complete for view {}!",
+            self.id, event, *view_number
+        );
+
         match event.as_ref() {
+            HotShotEvent::ProposeNow(view, proposal_dependency_data) => {
+                payload_commitment_dependency.mark_as_completed(
+                    HotShotEvent::SendPayloadCommitmentAndMetadata(
+                        proposal_dependency_data.commitment_and_metadata.commitment,
+                        proposal_dependency_data
+                            .commitment_and_metadata
+                            .metadata
+                            .clone(),
+                        *view,
+                    )
+                    .into(),
+                );
+
+                match &proposal_dependency_data.secondary_proposal_information {
+                    hotshot_types::consensus::SecondaryProposalInformation::QuorumProposalAndCertificate(quorum_proposal, leaf, qc) => {
+                        proposal_dependency.mark_as_completed(HotShotEvent::QuorumProposalValidated(quorum_proposal.clone(), leaf.clone()).into());
+                        qc_dependency.mark_as_completed(HotShotEvent::QCFormed(either::Left(qc.clone())).into());
+                    } ,
+                    hotshot_types::consensus::SecondaryProposalInformation::Timeout(tc) => {
+                        timeout_dependency.mark_as_completed(HotShotEvent::QCFormed(either::Right(tc.clone())).into());
+                    },
+                    hotshot_types::consensus::SecondaryProposalInformation::ViewSync(vs) => {
+                        view_sync_dependency.mark_as_completed(HotShotEvent::ViewSyncFinalizeCertificate2Recv(vs.clone()).into());
+                    },
+                };
+            }
```

Comment:
My understanding of this part is that it marks applicable dependencies as completed, which differ due to the SecondaryProposalInformation, but no matter which dependencies they are, we are basically completing sufficient dependencies so that we can send the proposal.

If my above understanding is correct, is this equivalent to defining the task completion as something like OrDependency(ProposalNow, CurrentDependencyLogic)? In this way, we don't need to do these mark_as_completed operations anymore.

Reply:
Oh that's a good point!